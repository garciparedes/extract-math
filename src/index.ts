import { default as escapeStringRegexp } from 'escape-string-regexp'

export interface Segment {
  type: 'text' | 'display' | 'inline'
  math: boolean
  value: string
  raw: string
}
const ESCAPE_DELIMITER = escapeStringRegexp('\\')

export function extractMath (input: string, inlineDelimiter: string = '$', displayDelimiter: string = '$$'): Segment[] {
  const segments: Segment[] = []
  const pattern: RegExp = buildRegExp(inlineDelimiter, displayDelimiter)

  let textWithInline: string
  let textWithDisplay: string
  let display: string
  let inline: string

  let [text, ...parts] = input.split(pattern)

  pushText(segments, text)

  while (parts.length > 0) {
    [textWithInline, textWithDisplay, display, inline, ...parts] = parts

    if (textWithInline) {
      pushText(segments, inlineDelimiter)
    } else if (textWithDisplay) {
      pushText(segments, displayDelimiter)
    } else if (display) {
      pushMath(segments, 'display', display, inlineDelimiter, displayDelimiter)
    } else if (inline) {
      pushMath(segments, 'inline', inline, inlineDelimiter, displayDelimiter)
    }

    [text, ...parts] = parts

    pushText(segments, text)
  }

  return segments
}

function buildRegExp (inlineDelimiter: string, displayDelimiter: string): RegExp {
  const textWithInline = buildTextWithInlineRegExp(inlineDelimiter)
  const textWithDisplay = buildTextWithDisplayRegExp(displayDelimiter)
  const display = buildDisplayRegExp(displayDelimiter)
  const inline = buildInlineRegExp(inlineDelimiter)

  return new RegExp([textWithInline.source, textWithDisplay.source, display.source, inline.source].join('|'))
}

function buildTextWithInlineRegExp (inlineDelimiter: string): RegExp {
  inlineDelimiter = escapeStringRegexp(inlineDelimiter)
  return new RegExp(`(${ESCAPE_DELIMITER}${inlineDelimiter})`)
}

function buildTextWithDisplayRegExp (displayDelimiter: string): RegExp {
  displayDelimiter = escapeStringRegexp(displayDelimiter)
  return new RegExp(`(${ESCAPE_DELIMITER}${displayDelimiter})`)
}

function buildDisplayRegExp (displayDelimiter: string): RegExp {
  displayDelimiter = escapeStringRegexp(displayDelimiter)
  return new RegExp(`${displayDelimiter}(.*?[^${ESCAPE_DELIMITER}])${displayDelimiter}`)
}

function buildInlineRegExp (inlineDelimiter: string): RegExp {
  inlineDelimiter = escapeStringRegexp(inlineDelimiter)
  return new RegExp(`${inlineDelimiter}(.*?[^${ESCAPE_DELIMITER}])${inlineDelimiter}`)
}

function pushText (segments: Segment[], text: string) {
  if (!text) {
    return
  }

  const last = segments[segments.length - 1]

  if (last && last.type === 'text') {
    last.value += text
    last.raw += text
  } else {
    segments.push({ type: 'text', math: false, value: text, raw: text })
  }
}

function pushMath (segments: Segment[], mode: 'inline' | 'display', text: string,
                   inlineDelimiter: string, displayDelimiter: string) {
  if (!text) {
    return
  }

  const value = cleanValue(text, inlineDelimiter, displayDelimiter)

  segments.push({ type: mode, math: true, value: value, raw: text })
}

function cleanValue (text: string, inlineDelimiter: string, displayDelimiter: string) {
  return text
    .replace(buildReplaceDisplayRegExp(displayDelimiter), displayDelimiter)
    .replace(buildReplaceInlineRegExp(inlineDelimiter), inlineDelimiter)
}

function buildReplaceDisplayRegExp (displayDelimiter: string) {
  displayDelimiter = escapeStringRegexp(displayDelimiter)
  return new RegExp(`${ESCAPE_DELIMITER}${displayDelimiter}`, 'g')
}

function buildReplaceInlineRegExp (inlineDelimiter: string) {
  inlineDelimiter = escapeStringRegexp(inlineDelimiter)
  return new RegExp(`${ESCAPE_DELIMITER}${inlineDelimiter}`, 'g')
}
