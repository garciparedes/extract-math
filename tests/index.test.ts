import { extractMath, Delimiters, LATEX_DELIMITERS } from '../src'

test('empty input', () => {
  const segments = extractMath('')
  expect(segments).toEqual([])
})

test('plain text', () => {
  const segments = extractMath('hello world')
  expect(segments).toEqual([
    {
      type: 'text',
      math: false,
      value: 'hello world',
      raw: 'hello world'
    }
  ])
})

test('escaped dollar', () => {
  const segments = extractMath('\\$')
  expect(segments).toEqual([
    {
      type: 'text',
      math: false,
      value: '$',
      raw: '$'
    }
  ])
})

test('inline math', () => {
  const segments = extractMath('$123 + \\$1$')
  expect(segments).toEqual([
    {
      type: 'inline',
      math: true,
      value: '123 + $1',
      raw: '123 + \\$1'
    }
  ])
})

test('display math', () => {
  const segments = extractMath('$$123 + \\$1$$')
  expect(segments).toEqual([
    {
      type: 'display',
      math: true,
      value: '123 + $1',
      raw: '123 + \\$1'
    }
  ])
})

test('combined text', () => {
  const segments = extractMath(`
    Text with an inline formula $x ^ 2 + 5$ and a displayed equation:

    $$\\sum_{i=1}^n(x_i^2 - \\overline{x}^2)$$

    The inline formula is represented as \\$expression\\$ and the displayed equation as \\$\\$expression\\$\\$.
    The \\$ symbol can be escaped like this: \\\\$
  `)
  expect(segments).toMatchSnapshot()
})

test('combined text custom delimiters', () => {
  const segments = extractMath(`
    Text with an inline formula @x ^ 2 + 5@ and a displayed equation:

    &\\sum_{i=1}^n(x_i^2 - \\overline{x}^2)&

    The inline formula is represented as \\@expression\\@ and the displayed equation as \\&expression\\&.
    The \\@ symbol can be escaped like this: \\\\@
    The \\& symbol can be escaped like this: \\\\&
  `, new Delimiters('@', '&'))
  expect(segments).toMatchSnapshot()
})

test('combined text custom opening and closing delimiters', () => {
  const segments = extractMath(`
    Text with an inline formula \\(x ^ 2 + 5\\) and a displayed equation:

    \\[\\sum_{i=1}^n(x_i^2 - \\overline{x}^2)\\]

    The inline formula is represented as \\\\(expression\\\\) and the displayed equation as \\\\[expression\\\\].
    The \\\\( symbol can be escaped like this: \\\\\\(
    The \\\\) symbol can be escaped like this: \\\\\\)
    The \\\\[ symbol can be escaped like this: \\\\\\[
    The \\\\] symbol can be escaped like this: \\\\\\]
  `, LATEX_DELIMITERS)
  expect(segments).toMatchSnapshot()
})

test('escaped dollar in math', () => {
  const segments = extractMath(`$$hello\\$$$$t\\$hing\\$$\\$\\$$a$a`)
  expect(segments).toEqual([
    { type: 'display', math: true, value: 'hello$', raw: 'hello\\$' },
    { type: 'inline', math: true, value: 't$hing$', raw: 't\\$hing\\$' },
    { type: 'text', math: false, value: '$$', raw: '$$' },
    { type: 'inline', math: true, value: 'a', raw: 'a' },
    { type: 'text', math: false, value: 'a', raw: 'a' }
  ])
})
