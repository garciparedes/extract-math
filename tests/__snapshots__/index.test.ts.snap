// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`combined text 1`] = `
Array [
  Object {
    "math": false,
    "raw": "
    Text with an inline formula ",
    "type": "text",
    "value": "
    Text with an inline formula ",
  },
  Object {
    "math": true,
    "raw": "x ^ 2 + 5",
    "type": "inline",
    "value": "x ^ 2 + 5",
  },
  Object {
    "math": false,
    "raw": " and a displayed equation:

    ",
    "type": "text",
    "value": " and a displayed equation:

    ",
  },
  Object {
    "math": true,
    "raw": "\\\\sum_{i=1}^n(x_i^2 - \\\\overline{x}^2)",
    "type": "display",
    "value": "\\\\sum_{i=1}^n(x_i^2 - \\\\overline{x}^2)",
  },
  Object {
    "math": false,
    "raw": "

    The inline formula is represented as $expression$ and the displayed equation as $$expression$$.
    The $ symbol can be escaped like this: \\\\$
  ",
    "type": "text",
    "value": "

    The inline formula is represented as $expression$ and the displayed equation as $$expression$$.
    The $ symbol can be escaped like this: \\\\$
  ",
  },
]
`;

exports[`combined text custom delimiters 1`] = `
Array [
  Object {
    "math": false,
    "raw": "
    Text with an inline formula ",
    "type": "text",
    "value": "
    Text with an inline formula ",
  },
  Object {
    "math": true,
    "raw": "x ^ 2 + 5",
    "type": "inline",
    "value": "x ^ 2 + 5",
  },
  Object {
    "math": false,
    "raw": " and a displayed equation:

    ",
    "type": "text",
    "value": " and a displayed equation:

    ",
  },
  Object {
    "math": true,
    "raw": "\\\\sum_{i=1}^n(x_i^2 - \\\\overline{x}^2)",
    "type": "display",
    "value": "\\\\sum_{i=1}^n(x_i^2 - \\\\overline{x}^2)",
  },
  Object {
    "math": false,
    "raw": "

    The inline formula is represented as @expression@ and the displayed equation as &expression&.
    The @ symbol can be escaped like this: \\\\@
    The & symbol can be escaped like this: \\\\&
  ",
    "type": "text",
    "value": "

    The inline formula is represented as @expression@ and the displayed equation as &expression&.
    The @ symbol can be escaped like this: \\\\@
    The & symbol can be escaped like this: \\\\&
  ",
  },
]
`;

exports[`combined text custom opening and closing delimiters 1`] = `
Array [
  Object {
    "math": false,
    "raw": "
    Text with an inline formula ",
    "type": "text",
    "value": "
    Text with an inline formula ",
  },
  Object {
    "math": true,
    "raw": "x ^ 2 + 5",
    "type": "inline",
    "value": "x ^ 2 + 5",
  },
  Object {
    "math": false,
    "raw": " and a displayed equation:

    ",
    "type": "text",
    "value": " and a displayed equation:

    ",
  },
  Object {
    "math": true,
    "raw": "\\\\sum_{i=1}^n(x_i^2 - \\\\overline{x}^2)",
    "type": "display",
    "value": "\\\\sum_{i=1}^n(x_i^2 - \\\\overline{x}^2)",
  },
  Object {
    "math": false,
    "raw": "

    The inline formula is represented as \\\\(expression\\\\) and the displayed equation as \\\\[expression\\\\].
    The \\\\( symbol can be escaped like this: \\\\\\\\(
    The \\\\) symbol can be escaped like this: \\\\\\\\)
    The \\\\[ symbol can be escaped like this: \\\\\\\\[
    The \\\\] symbol can be escaped like this: \\\\\\\\]
  ",
    "type": "text",
    "value": "

    The inline formula is represented as \\\\(expression\\\\) and the displayed equation as \\\\[expression\\\\].
    The \\\\( symbol can be escaped like this: \\\\\\\\(
    The \\\\) symbol can be escaped like this: \\\\\\\\)
    The \\\\[ symbol can be escaped like this: \\\\\\\\[
    The \\\\] symbol can be escaped like this: \\\\\\\\]
  ",
  },
]
`;
